### ISSUE TITLE
v1.1-beta: A new moved block feature doesn't check resource type

### ISSUE THREAD

**Issue by minamijoyo** [hashicorp/terraform#29896]
Hi there, I found a bug for a new `moved` block feature in Terraform v1.1.0-beta1.

### Terraform Version

```
$ terraform -v
Terraform v1.1.0-beta1
on darwin_amd64
+ provider registry.terraform.io/hashicorp/null v3.1.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
```

### Terraform Configuration Files

Given the following configuration before moving

```main.tf
resource "null_resource" "foo1" {}
resource "null_resource" "foo2" {}
resource "random_uuid" "foo3" {}
```

Then, rename `null_resource.foo1` to `random_uuid.foo1`

```main.tf
resource "random_uuid" "foo1" {}
resource "null_resource" "foo2" {}
resource "random_uuid" "foo3" {}

moved {
  from = null_resource.foo1
  to   = random_uuid.foo1
}
```

### Expected Behavior

As described in doc, we cannot change a resource type by using `moved` block.

https://github.com/hashicorp/terraform/blob/v1.1.0-beta1/website/docs/language/modules/develop/refactoring.html.md#renaming-a-resource

> Each resource type has a separate schema and so objects of different types are not compatible. Therefore, although you can use moved to change the name of a resource, you cannot use moved to change to a different resource type or to change a managed resource (a resource block) into a data resource (a data block).

### Actual Behavior

We can actually change the resource type by using `moved` block and it results in a broken state.

```
$ terraform apply -auto-approve
$ terraform state list
null_resource.foo2
random_uuid.foo1
random_uuid.foo3
```

```
$ terraform state pull
{
  "version": 4,
  "terraform_version": "1.1.0",
  "serial": 4,
  "lineage": "6f3db6e1-d10a-bbd4-8203-b5e36753b84b",
  "outputs": {},
  "resources": [
    {
      "mode": "managed",
      "type": "null_resource",
      "name": "foo2",
      "provider": "provider[\"registry.terraform.io/hashicorp/null\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "4495596860553745594",
            "triggers": null
          },
          "sensitive_attributes": [],
          "private": "bnVsbA=="
        }
      ]
    },
    {
      "mode": "managed",
      "type": "random_uuid",
      "name": "foo1",
      "provider": "provider[\"registry.terraform.io/hashicorp/random\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "1447943017614191348",
            "keepers": null,
            "result": null
          },
          "sensitive_attributes": [],
          "private": "bnVsbA=="
        }
      ]
    },
    {
      "mode": "managed",
      "type": "random_uuid",
      "name": "foo3",
      "provider": "provider[\"registry.terraform.io/hashicorp/random\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "06c1f9cd-8db0-7403-0786-461f46f205f6",
            "keepers": null,
            "result": "06c1f9cd-8db0-7403-0786-461f46f205f6"
          },
          "sensitive_attributes": [],
          "private": "bnVsbA=="
        }
      ]
    }
  ]
}
```

Note that the `id` of `random_uuid.foo1` is not in uuid format.

### Steps to Reproduce

1. Create the following configuration and apply it:

```
$ cat main.tf
resource "null_resource" "foo1" {}
resource "null_resource" "foo2" {}
resource "random_uuid" "foo3" {}

$ terraform apply -auto-approve

$ terraform state list
null_resource.foo1
null_resource.foo2
random_uuid.foo3
```

2. First, try to rename `null_resource.foo1` to `random_uuid.foo1` with the `state mv` command. This fails as expected:

```
$ terraform state mv null_resource.foo1 random_uuid.foo1
╷
│ Error: Invalid state move request
│
│ Cannot move null_resource.foo1 to random_uuid.foo1: resource types don't match.
╵

$ echo $?
1

$ terraform state list
null_resource.foo1
null_resource.foo2
random_uuid.foo3
```

3. Let's rename `null_resource.foo1` to `random_uuid.foo1` with the `moved` block:

```
$ cat main.tf
resource "random_uuid" "foo1" {}
resource "null_resource" "foo2" {}
resource "random_uuid" "foo3" {}

moved {
  from = null_resource.foo1
  to   = random_uuid.foo1
}

$ terraform validate
Success! The configuration is valid.

$ terraform plan
null_resource.foo2: Refreshing state... [id=4495596860553745594]
random_uuid.foo1: Refreshing state... [id=1447943017614191348]
random_uuid.foo3: Refreshing state... [id=06c1f9cd-8db0-7403-0786-461f46f205f6]

Terraform will perform the following actions:

  # null_resource.foo1 has moved to random_uuid.foo1
    resource "random_uuid" "foo1" {
        id = "1447943017614191348"
    }

Plan: 0 to add, 0 to change, 0 to destroy.

──────────────────────────────────────────────────────────────────────────────────────────────────────

Note: You didn't use the -out option to save this plan, so Terraform can't guarantee to take exactly
these actions if you run "terraform apply" now.

$ terraform apply -auto-approve
random_uuid.foo3: Refreshing state... [id=06c1f9cd-8db0-7403-0786-461f46f205f6]
random_uuid.foo1: Refreshing state... [id=1447943017614191348]
null_resource.foo2: Refreshing state... [id=4495596860553745594]

Terraform will perform the following actions:

  # null_resource.foo1 has moved to random_uuid.foo1
    resource "random_uuid" "foo1" {
        id = "1447943017614191348"
    }

Plan: 0 to add, 0 to change, 0 to destroy.

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

$ terraform state list
null_resource.foo2
random_uuid.foo1
random_uuid.foo3

$ terraform state pull
{
  "version": 4,
  "terraform_version": "1.1.0",
  "serial": 4,
  "lineage": "6f3db6e1-d10a-bbd4-8203-b5e36753b84b",
  "outputs": {},
  "resources": [
    {
      "mode": "managed",
      "type": "null_resource",
      "name": "foo2",
      "provider": "provider[\"registry.terraform.io/hashicorp/null\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "4495596860553745594",
            "triggers": null
          },
          "sensitive_attributes": [],
          "private": "bnVsbA=="
        }
      ]
    },
    {
      "mode": "managed",
      "type": "random_uuid",
      "name": "foo1",
      "provider": "provider[\"registry.terraform.io/hashicorp/random\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "1447943017614191348",
            "keepers": null,
            "result": null
          },
          "sensitive_attributes": [],
          "private": "bnVsbA=="
        }
      ]
    },
    {
      "mode": "managed",
      "type": "random_uuid",
      "name": "foo3",
      "provider": "provider[\"registry.terraform.io/hashicorp/random\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "06c1f9cd-8db0-7403-0786-461f46f205f6",
            "keepers": null,
            "result": "06c1f9cd-8db0-7403-0786-461f46f205f6"
          },
          "sensitive_attributes": [],
          "private": "bnVsbA=="
        }
      ]
    }
  ]
}
```

### Additional Context

I happened to find this bug when testing the beta build. I don't have a specific use case for it. However, this behavior might make sense in some cases, for examples, changing a deprecated resource type to new one. If we allow it intentionally, we should warn it as unsafe.

**Comment by apparentlymart** [comment]
Thanks for reporting this, @minamijoyo!

You're correct that our design intent is for Terraform to catch and report this as an error. I think perhaps what happened here is that an earlier iteration of this implementation did have the beginnings of a mechanism to migrate away from deprecated resource types (as you mentioned), but the design for that required additions to the provider protocol that we weren't able to finalize in time and so we deferred it for later work. My guess is that we removed that mechanism but forgot to replace it with a hard error rejecting resource type changes altogether.

We'll get this fixed up before final release. Thanks again!

**Comment by minamijoyo** [comment]
@apparentlymart Thank you for your reply and the explanation of context. As you said, returning the error for now sounds reasonable for me too.

**Comment by minamijoyo** [comment]
@kmoe Thank you for the fix! I confirmed that it was fixed in v1.1-beta2 🎉

**Comment by apparentlymart** [comment]
Thanks for confirming, @minamijoyo! (and again, for reporting it in the first place!)
